src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.AddSub.cs

```diff
@@ -20,50 +21,50 @@ public static uint[] Add(uint[] left, uint right)
 
             uint[] bits = new uint[left.Length + 1];
 
-            long digit = (long)left[0] + right;
-            bits[0] = unchecked((uint)digit);
+            Add(left, right, ref GetArrayDataReference(bits));
+
+            return bits;
+        }
+
+        private static void Add(ReadOnlySpan<uint> left,
+                                uint right,
+                                ref uint bits)
+        {
+            long digit = (long)GetReference(left) + right;
+            bits = unchecked((uint)digit);
             long carry = digit >> 32;
 
             for (int i = 1; i < left.Length; i++)
             {
-                digit = left[i] + carry;
-                bits[i] = unchecked((uint)digit);
+                digit = Unsafe.Add(ref GetReference(left), i) + carry;
```

Comment:
```suggestion
                digit = left[i] + carry;
```
No need for `Unsafe`. If the bound-check matters, one could slice `left` so the for-loop iterates from `0` to `left.Length`, and the JIT can elide the bound-checks and we get safety and more readable code back.

Same at other places.